File: backend/app/schemas/ai_tests.py
"""Schemas for AI test generation, validation, and results."""
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.schemas.test_execution import (
    ExceptionType,
    FunctionCombinedTestResult,
    FunctionTestCase,
    ObjectScenarioTestCase,
    ObjectScenarioTestResult,
    TemplateArgumentInput,
)


# ---------------------------------------------------------------------------
# Error class
# ---------------------------------------------------------------------------


class AiTestServiceError(RuntimeError):
    def __init__(self, code: str, message: str, status_code: int = 502):
        super().__init__(message)
        self.code = code
        self.message = message
        self.status_code = status_code


# ---------------------------------------------------------------------------
# Coverage categories
# ---------------------------------------------------------------------------

AiCoverageCategory = Literal[
    "normal",
    "zero",
    "negative",
    "boundary",
    "below_boundary",
    "above_boundary",
    "empty",
    "single_element",
    "duplicate",
    "ordering",
    "already_sorted",
    "reverse_sorted",
    "not_found",
    "first_element",
    "last_element",
    "mutation",
    "exception",
    "iterator_begin",
    "iterator_end",
    "empty_range",
    "full_range",
    "partial_range",
    "object_state",
    "template_case",
]

# ---------------------------------------------------------------------------
# Gemini response schemas — strict, extra="forbid" on every model
# ---------------------------------------------------------------------------

_STRICT = ConfigDict(extra="forbid")

_CPP_KEYWORDS = frozenset(
    {
        "alignas", "alignof", "and", "and_eq", "asm", "auto", "bitand", "bitor",
        "bool", "break", "case", "catch", "char", "char8_t", "char16_t", "char32_t",
        "class", "compl", "concept", "const", "consteval", "constexpr", "constinit",
        "const_cast", "continue", "co_await", "co_return", "co_yield", "decltype",
        "default", "delete", "do", "double", "dynamic_cast", "else", "enum",
        "explicit", "export", "extern", "false", "float", "for", "friend", "goto",
        "if", "inline", "int", "long", "mutable", "namespace", "new", "noexcept",
        "not", "not_eq", "nullptr", "operator", "or", "or_eq", "private", "protected",
        "public", "register", "reinterpret_cast", "requires", "return", "short",
        "signed", "sizeof", "static", "static_assert", "static_cast", "struct",
        "switch", "template", "this", "thread_local", "throw", "true", "try",
        "typedef", "typeid", "typename", "union", "unsigned", "using", "virtual",
        "void", "volatile", "wchar_t", "while", "xor", "xor_eq",
    }
)


class AiModelMutation(BaseModel):
    model_config = _STRICT
    parameter_name: str = Field(min_length=1, max_length=100)
    expected_final_value: str = Field(max_length=1000)


class AiModelFunctionTest(BaseModel):
    model_config = _STRICT
    name: str = Field(min_length=1, max_length=80)
    category: AiCoverageCategory
    reason: str = Field(min_length=1, max_length=200)
    arguments: list[str] = Field(max_length=20)
    expected_outcome: Literal["return_value", "return_void", "throws"]
    expected_return: str | None = Field(default=None, max_length=1000)
    expected_stdout: str | None = Field(default=None, max_length=2000)
    expected_mutations: list[AiModelMutation] = Field(
        default_factory=list, max_length=20
    )
    expected_exception_type: ExceptionType | None = None
    exception_message_rule: Literal["ignore", "contains"] = "ignore"
    expected_exception_message: str | None = Field(default=None, max_length=200)


class AiModelTestPlan(BaseModel):
    model_config = _STRICT
    target_id: str = Field(min_length=1, max_length=300)
    tests: list[AiModelFunctionTest] = Field(min_length=0, max_length=8)
    skipped_topics: list[str] = Field(default_factory=list, max_length=5)

    @field_validator("skipped_topics")
    @classmethod
    def _validate_skipped(cls, v: list[str]) -> list[str]:
        for item in v:
            if len(item) > 200:
                raise ValueError("Each skipped_topics entry must be ≤ 200 characters.")
        return v


# --- Object variant ---


class AiModelScenarioObject(BaseModel):
    model_config = _STRICT
    name: str = Field(min_length=1, max_length=50)
    constructor_id: str
    arguments: list[str] = Field(max_length=20)
    expected_outcome: Literal["return_void", "throws"] = "return_void"
    expected_exception_type: ExceptionType | None = None

    @field_validator("name")
    @classmethod
    def _no_keywords(cls, v: str) -> str:
        if not v.isidentifier():
            raise ValueError(f"Object name '{v}' is not a valid C++ identifier.")
        if v in _CPP_KEYWORDS:
            raise ValueError(f"Object name '{v}' collides with a C++ keyword.")
        return v


class AiModelScenarioStep(BaseModel):
    model_config = _STRICT
    step_type: Literal["method", "observer"]
    target_object_name: str
    method_id: str
    arguments: list[str] = Field(default_factory=list, max_length=20)
    expected_outcome: Literal["return_value", "return_void", "throws"]
    expected_return: str | None = Field(default=None, max_length=1000)
    check_stdout: bool = False
    expected_stdout: str | None = Field(default=None, max_length=2000)
    expected_exception_type: ExceptionType | None = None
    exception_message_rule: Literal["ignore", "contains"] = "ignore"
    expected_exception_message: str | None = Field(default=None, max_length=200)


class AiModelScenarioTest(BaseModel):
    model_config = _STRICT
    name: str = Field(min_length=1, max_length=80)
    category: AiCoverageCategory
    reason: str = Field(min_length=1, max_length=200)
    objects: list[AiModelScenarioObject] = Field(min_length=1, max_length=5)
    steps: list[AiModelScenarioStep] = Field(min_length=1, max_length=10)


class AiModelScenarioPlan(BaseModel):
    model_config = _STRICT
    target_id: str = Field(min_length=1, max_length=300)
    tests: list[AiModelScenarioTest] = Field(min_length=0, max_length=8)
    skipped_topics: list[str] = Field(default_factory=list, max_length=5)

    @field_validator("skipped_topics")
    @classmethod
    def _validate_skipped(cls, v: list[str]) -> list[str]:
        for item in v:
            if len(item) > 200:
                raise ValueError("Each skipped_topics entry must be ≤ 200 characters.")
        return v


# ---------------------------------------------------------------------------
# API request/response models
# ---------------------------------------------------------------------------


class AiScore(BaseModel):
    passed: int
    executed: int
    percentage: int


class AiTestResultRow(BaseModel):
    id: str
    name: str
    category: AiCoverageCategory
    reason: str
    passed: bool
    input_summary: str
    expected_summary: str
    actual_summary: str
    detail: FunctionCombinedTestResult | ObjectScenarioTestResult | None = None


class AiStoredTest(BaseModel):
    id: str
    name: str
    category: AiCoverageCategory
    reason: str
    function_test: FunctionTestCase | None = None
    scenario_test: ObjectScenarioTestCase | None = None

    @model_validator(mode="after")
    def _exactly_one(self) -> "AiStoredTest":
        has_fn = self.function_test is not None
        has_sc = self.scenario_test is not None
        if has_fn == has_sc:
            raise ValueError(
                "Exactly one of function_test or scenario_test must be present."
            )
        return self


_AI_STATUS = Literal[
    "completed",
    "missing_question",
    "compile_failed",
    "unsupported",
    "source_too_large",
    "generation_timeout",
    "generation_rate_limited",
    "generation_unavailable",
    "generation_failed",
    "no_useful_tests",
    "infrastructure_failed",
    "incompatible",
]

PRACTICE_DISCLAIMER = (
    "This is a practice score based on AI-generated tests, "
    "not an official course grade."
)


class AiTestRunResponse(BaseModel):
    status: _AI_STATUS
    message: str | None = None
    unsupported_reason: str | None = None
    unsupported_parameters: list[list[str]] = []
    supported_targets: list[str] = []
    score: AiScore | None = None
    tests: list[AiTestResultRow] = []
    stored_tests: list[AiStoredTest] = []
    skipped_topics: list[str] = []
    generation_note: str | None = None
    memory_status: Literal["not_run"] = "not_run"
    disclaimer: str = PRACTICE_DISCLAIMER


class AiTestRunRequest(BaseModel):
    code: str = Field(min_length=1, max_length=1_000_000)
    language: Literal["cpp"]
    question_text: str = Field(min_length=1, max_length=8000)
    target_kind: Literal["function", "object"]
    target_id: str = Field(min_length=1, max_length=300)
    template_argument_mode: Literal["deduced", "explicit"] | None = None
    template_arguments: list[TemplateArgumentInput] = Field(
        default_factory=list, max_length=10
    )


class AiTestRerunRequest(BaseModel):
    code: str = Field(min_length=1, max_length=1_000_000)
    language: Literal["cpp"]
    target_kind: Literal["function", "object"]
    target_id: str = Field(min_length=1, max_length=300)
    template_argument_mode: Literal["deduced", "explicit"] | None = None
    template_arguments: list[TemplateArgumentInput] = Field(
        default_factory=list, max_length=10
    )
    tests: list[AiStoredTest] = Field(min_length=1, max_length=8)

File: frontend/lib/aiTests.ts
/**
 * API helpers for the AI test generation endpoints.
 * Wire-type mirrors of the backend schemas (ai_tests.py).
 */

// ---------------------------------------------------------------------------
// Wire types (mirrors of backend schemas)
// ---------------------------------------------------------------------------

export type AiScore = {
  passed: number;
  executed: number;
  percentage: number;
};

export type AiTestResultRow = {
  id: string;
  name: string;
  category: string;
  reason: string;
  passed: boolean;
  input_summary: string;
  expected_summary: string;
  actual_summary: string;
  detail: unknown | null;
};

export type AiStoredTest = {
  id: string;
  name: string;
  category: string;
  reason: string;
  function_test?: unknown;
  scenario_test?: unknown;
};

export type TemplateArgumentInput = {
  parameter_name: string;
  kind: "type" | "non_type";
  value: string;
};

export type AiTestRunRequest = {
  code: string;
  language: "cpp";
  question_text: string;
  target_kind: "function" | "object";
  target_id: string;
  template_argument_mode?: "deduced" | "explicit" | null;
  template_arguments?: TemplateArgumentInput[];
};

export type AiTestRerunRequest = {
  code: string;
  language: "cpp";
  target_kind: "function" | "object";
  target_id: string;
  template_argument_mode?: "deduced" | "explicit" | null;
  template_arguments?: TemplateArgumentInput[];
  tests: AiStoredTest[];
};

export type AiTestRunResponseStatus =
  | "completed"
  | "missing_question"
  | "compile_failed"
  | "unsupported"
  | "source_too_large"
  | "generation_timeout"
  | "generation_rate_limited"
  | "generation_unavailable"
  | "generation_failed"
  | "no_useful_tests"
  | "infrastructure_failed"
  | "incompatible";

export type AiTestRunResponse = {
  status: AiTestRunResponseStatus;
  message: string | null;
  unsupported_reason: string | null;
  unsupported_parameters: [string, string][];
  supported_targets: string[];
  score: AiScore | null;
  tests: AiTestResultRow[];
  stored_tests: AiStoredTest[];
  skipped_topics: string[];
  generation_note: string | null;
  memory_status: "not_run";
  disclaimer: string;
};

export type QuestionTextResponse = {
  question_text: string;
  model: string;
};

// ---------------------------------------------------------------------------
// Error class
// ---------------------------------------------------------------------------

export class AiTestsRequestError extends Error {
  constructor(
    message: string,
    public readonly status: AiTestRunResponseStatus | null,
  ) {
    super(message);
    this.name = "AiTestsRequestError";
  }
}

// ---------------------------------------------------------------------------
// API base URL
// ---------------------------------------------------------------------------

function apiBase(): string {
  return process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://localhost:8000";
}

type ApiError = { error?: { message?: string } };

// ---------------------------------------------------------------------------
// Run AI tests
// ---------------------------------------------------------------------------

export async function runAiTests(
  request: AiTestRunRequest,
): Promise<AiTestRunResponse> {
  try {
    const response = await fetch(`${apiBase()}/api/ai-tests/run`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(request),
    });
    const body: unknown = await response.json().catch(() => null);

    if (!response.ok) {
      const apiError = body as ApiError | null;
      throw new AiTestsRequestError(
        apiError?.error?.message ?? "AI test generation failed. Please retry.",
        null,
      );
    }
    return body as AiTestRunResponse;
  } catch (error) {
    if (error instanceof AiTestsRequestError) throw error;
    throw new AiTestsRequestError(
      "The AI test service is unavailable. Check the backend and retry.",
      null,
    );
  }
}

// ---------------------------------------------------------------------------
// Rerun AI tests
// ---------------------------------------------------------------------------

export async function rerunAiTests(
  request: AiTestRerunRequest,
): Promise<AiTestRunResponse> {
  try {
    const response = await fetch(`${apiBase()}/api/ai-tests/rerun`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(request),
    });
    const body: unknown = await response.json().catch(() => null);

    if (!response.ok) {
      const apiError = body as ApiError | null;
      throw new AiTestsRequestError(
        apiError?.error?.message ?? "AI test rerun failed. Please retry.",
        null,
      );
    }
    return body as AiTestRunResponse;
  } catch (error) {
    if (error instanceof AiTestsRequestError) throw error;
    throw new AiTestsRequestError(
      "The AI test service is unavailable. Check the backend and retry.",
      null,
    );
  }
}

// ---------------------------------------------------------------------------
// Transcribe question pages
// ---------------------------------------------------------------------------

export async function transcribeQuestion(
  formData: FormData,
): Promise<QuestionTextResponse> {
  try {
    const response = await fetch(`${apiBase()}/api/transcribe-question`, {
      method: "POST",
      body: formData,
    });
    const body: unknown = await response.json().catch(() => null);

    if (!response.ok) {
      const apiError = body as ApiError | null;
      throw new AiTestsRequestError(
        apiError?.error?.message ?? "Question extraction failed. Please retry.",
        null,
      );
    }
    return body as QuestionTextResponse;
  } catch (error) {
    if (error instanceof AiTestsRequestError) throw error;
    throw new AiTestsRequestError(
      "The question extraction service is unavailable.",
      null,
    );
  }
}
